File: problems/0076-minimum-window-substring/solution.py
from collections import Counter

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        """
        Problem: 0076 - Minimum Window Substring

        Given two strings s and t, return the minimum window in s which contains all the characters in t.
        If there is no such window, return the empty string.

        Approach:
        - Use the sliding window technique with two pointers (left, right).
        - Count the frequency of characters needed from t using a hash map.
        - Expand the right pointer to include more characters until the window is valid.
        - Then, move the left pointer to shrink the window while still maintaining validity.
        - Keep track of the minimum valid window seen.

        Time complexity: O(m + n), where m = len(s), n = len(t).
        Space complexity: O(n) for the character counts.
        """
        if not t or not s:
            return ""

        t_count = Counter(t)  # Frequency of required characters.
        window = {}  # Frequency of current window characters.
        have, need = 0, len(t_count)

        res = [-1, -1]  # Start and end of best window.
        res_len = float("inf")
        left = 0

        # Expand the window with the right pointer.
        for right in range(len(s)):
            char = s[right]
            window[char] = window.get(char, 0) + 1

            # Check if this character satisfies one required frequency.
            if char in t_count and window[char] == t_count[char]:
                have += 1

            # Try to contract the window from the left.
            while have == need:
                # Update result if this window is smaller.
                if (right - left + 1) < res_len:
                    res = [left, right]
                    res_len = right - left + 1

                # Remove the leftmost character from the window.
                window[s[left]] -= 1
                if s[left] in t_count and window[s[left]] < t_count[s[left]]:
                    have -= 1  # No longer satisfies required count.
                left += 1  # Shrink the window.

        l, r = res
        return s[l:r+1] if res_len != float("inf") else ""

File: problems/0076-minimum-window-substring/solution.cpp
class Solution {
public:
    /*
    Problem: 0076 - Minimum Window Substring

    Given two strings s and t, return the minimum window in s which contains all the characters in t.
    If there is no such window, return the empty string.

    Approach:
    - Use the sliding window technique with two pointers (left and right).
    - Count the required frequency of characters in t using an unordered_map.
    - Expand the window by moving right to include more characters.
    - When the window contains all required characters, try to shrink it from the left.
    - Track the smallest valid window found during the process.

    Time Complexity: O(m + n), where m = length of s and n = length of t.
    Space Complexity: O(n) for character counts.
    */
    string minWindow(string s, string t) {
        if (t.empty() || s.empty()) {
            return "";  // Edge case: empty input.
        }

        unordered_map<char, int> t_count;
        for (char c : t) {
            t_count[c]++;  // Count required characters from t.
        }

        unordered_map<char, int> window;
        int have = 0;
        int need = t_count.size();

        int res_len = INT_MAX;
        int res_left = 0;
        int left = 0;

        // Expand the window with the right pointer.
        for (int right = 0; right < s.size(); ++right) {
            char c = s[right];
            window[c]++;

            // Check if current character meets required frequency.
            if (t_count.count(c) && window[c] == t_count[c]) {
                have++;
            }

            // Shrink the window from the left while it's valid.
            while (have == need) {
                // Update result if the current window is smaller.
                if ((right - left + 1) < res_len) {
                    res_left = left;
                    res_len = right - left + 1;
                }

                // Remove the leftmost character from the window.
                window[s[left]]--;
                if (t_count.count(s[left]) && window[s[left]] < t_count[s[left]]) {
                    have--;  // Lost a required character.
                }
                left++;  // Move left to shrink the window.
            }
        }

        return res_len == INT_MAX ? "" : s.substr(res_left, res_len);
    }
};

